server/remote_cache/action_cache_server/action_cache_server.go

```diff
@@ -242,3 +244,58 @@ func (s *ActionCacheServer) UpdateActionResult(ctx context.Context, req *repb.Up
 	}
 	return req.ActionResult, nil
 }
+
+// Inlines the contents of output files requested to be inlined as long as the
+// total size of the ActionResult is below maxResultSize.
+func (s *ActionCacheServer) maybeInlineOutputFiles(ctx context.Context, req *repb.GetActionResultRequest, ar *repb.ActionResult, maxResultSize int) {
+	if ar == nil || len(req.InlineOutputFiles) == 0 {
+		return
+	}
+	requestedFiles := make(map[string]struct{}, len(req.InlineOutputFiles))
+	for _, f := range req.InlineOutputFiles {
+		requestedFiles[f] = struct{}{}
+	}
+
+	budget := int64(max(0, maxResultSize-proto.Size(ar)))
+	var actuallyInlinedFiles []*repb.OutputFile
+	for _, f := range ar.OutputFiles {
+		if _, ok := requestedFiles[f.Path]; !ok {
+			continue
+		}
+		contentsSize := f.GetDigest().GetSizeBytes()
+		if contentsSize == 0 {
+			// Empty files don't need to be inlined as they can be recognized
+			// by their size.
+			continue
+		}
+		// An additional "contents" field requires 1 byte for the tag field
+		// (5:LEN), the bytes for the varint encoding of the length of the
+		// contents and the contents themselves.
+		totalSize := 1 + int64(binary.Size(uint64(contentsSize))) + contentsSize
+		if budget < totalSize {
+			continue
+		}
+		budget -= totalSize
+		actuallyInlinedFiles = append(actuallyInlinedFiles, f)
+	}
+
+	if len(actuallyInlinedFiles) == 0 {
+		return
+	}
+
+	g, gCtx := errgroup.WithContext(ctx)
+	for _, f := range actuallyInlinedFiles {
+		fc := f
+		g.Go(func() error {
+			rn := digest.NewResourceName(fc.GetDigest(), req.GetInstanceName(), rspb.CacheType_CAS, req.GetDigestFunction()).ToProto()
+			blob, err := s.cache.Get(gCtx, rn)
+			if err != nil {
+				// Inlining is best-effort ("The server MAY omit inlining...").
+				return nil
```

Comment:
Small nit on the error handling here (which should still apply when we switch to GetMulti): in the ValidateActionResult step right before this, we made sure that all the blobs referenced by the ActionResult are available before returning the action result. Else, we just return a not found error so Bazel could re-run the action.

The same principle should be applied here: if the referenced blob cannot be found in our cache, we should just return not found rather than try to not inline it. Because in subsequent requests from Bazel to download the blob, it might error out with cryptic message on the client side.

Bonus if we can combine some of these inline logic with the checkFilesExist in ValidateActionResult somehow (i.e. no need to validate blobs which we gonna inline). But it's probably not needed in the first iteration.